core/templates/dev/head/components/question-directives/question-player/question-player.directive.ts

```diff
@@ -368,6 +398,80 @@ angular.module('oppia').directive('questionPlayer', [
             $scope.resultsLoaded = true;
           };
 
+          var getMasteryChangeForWrongAnswers = function(
+              answers, masteryChangePerQuestion) {
+            answers.forEach(function(answer) {
+              if (!answer.isCorrect) {
+                if (answer.taggedSkillMisconceptionId) {
+                  var skillId = answer.taggedSkillMisconceptionId.split('-')[0];
+                  if (masteryChangePerQuestion.hasOwnProperty(skillId)) {
+                    masteryChangePerQuestion[skillId] -=
+                      WRONG_ANSWER_PENALTY_FOR_MASTERY;
+                  }
+                } else {
+                  for (var skillId in masteryChangePerQuestion) {
+                    masteryChangePerQuestion[skillId] -=
+                      WRONG_ANSWER_PENALTY_FOR_MASTERY;
+                  }
+                }
+              }
+            });
+            return masteryChangePerQuestion;
+          };
+
+          var updateMasteryPerSkillMapping = function(
+              masteryChangePerQuestion) {
+            for (var skillId in masteryChangePerQuestion) {
+              if (!(skillId in ctrl.masteryPerSkillMapping)) {
+                continue;
+              }
+              // Set the lowest bound of mastery change for each question.
+              ctrl.masteryPerSkillMapping[skillId] += Math.max(
+                masteryChangePerQuestion[skillId],
+                MAX_MASTERY_LOSS_PER_QUESTION);
+            }
+          };
+
+          var calculateMasteryDegrees = function(questionStateData) {
+            createMasteryPerSkillMapping();
+
+            for (var question in questionStateData) {
+              var questionData = questionStateData[question];
+              if (!(questionData.linkedSkillIds)) {
+                continue;
+              }
+              var masteryChangePerQuestion = {};
+              for (var i = 0; i < questionData.linkedSkillIds.length; i++) {
```

Comment:
Separate this into a function, something like 'initializeMasteryChangePerQuestion

Reply:
Done.